docs/source/scripts/attention-single-gpu.py

```diff
@@ -0,0 +1,348 @@
+# as the IR, so the traced module will become a ``torch.fx.GraphModule``, and we can also
+# see the forward function of it.
+
+print(attn_sch.mod)
+
+# %%
+# Second, the ``.find()`` primitive will return a list of subgraphs that
+# match the pattern. In our case, there will be three subgraphs, one for each
+# linear layer and the consequential ``view`` and ``permute`` operations.
+
+print(qkv_subgraphs)
+
+# %%
+# Then, we define a fused QKV module as follows and instantiate it.
+
+
+class FusedQKV(nn.Module):
+    def __init__(self, hidden_size, n_heads) -> None:
+        super().__init__()
+        self.n_heads = n_heads
+        self.fused_linear = nn.Linear(hidden_size, hidden_size * 3)
+
+    def permute_for_scores(self, x):
+        new_shape = x.shape[:-1] + (self.n_heads, -1)
+        x = x.view(new_shape)
+        return x.permute(0, 2, 1, 3)
+
+    def forward(self, hidden_states):
+        qkv = self.fused_linear(hidden_states)
+        reshaped_qkv = self.permute_for_scores(qkv)
+        q, k, v = torch.split(reshaped_qkv, 1, dim=-1)
+        q = torch.squeeze(q, -1).contiguous()
+        k = torch.squeeze(k, -1).contiguous()
+        v = torch.squeeze(v, -1).contiguous()
+        return [q, k, v]
+
+
+fused_qkv = FusedQKV(hidden_size=1024, n_heads=16)
+
+# %%
+# We can replace the subgraphs with the fused QKV module by calling the
+# ``.replace()`` primitive. The first argument is the new module,
+# and the second argument is the subgraph to be replaced.
+# After replacing the subgraph, we can check the model again to see the
+# changes.
+
+attn_sch.replace(fused_qkv, qkv_subgraphs)
+print(attn_sch.mod)
+
+# %%
+# From the above output, we can see there is a new module called ``FusedQKV_0``
+# with :math:`3\times` ``out_features`` compared to the original linear layer.
+# The corresponding forward function is also changed to leverage the fused
+# module.
+
+# %%
+# Replace Scaled Dot-Product Attention
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# Similarly, we can directly find the core attention function and replace it
+# with a more efficient implementation.
+
+core_attn_subgraph = attn_sch.find(scaled_dot_product)
+print(core_attn_subgraph)
+
+# %%
+# We can use the ``FlashAttentionOp`` from the ``xformers`` library to replace
+# the core attention, which has been included in Slapo, so we can directly
+# import and replace the subgraph with ``FlashAttentionOp``.
+# Notice, since the ``scaled_dot_product`` function we defined above only accepts
+# the ``query``, ``key``, and ``value`` tensors, while ``FlashAttentionOp`` requires
+# five arguments, so we need to explicitly pass ``None`` to the ``attention_mask``
+# argument, and set the dropout probability ``p`` to 0.1 by setting the ``concrete_args``.
+
+from slapo.op.attention import FlashAttentionOp
+
+flash_attn = FlashAttentionOp(attn_op_name="native_xformers", apply_causal_mask=False)
```

Comment:
Make a note saying that we use `native_xformers` in this tutorial to demonstrate the functionality. In reality, users could choose `cutlass`, `triton`, or `cuda` kernel; while the latter two only supports NVIDIA A100.